src/DistributedMetadata/CatalogService.cpp

```diff
@@ -49,18 +49,21 @@ const String CATALOG_DEFAULT_TOPIC = "__system_catalogs";
 
 const String THIS_HOST = getFQDNOrHostName();
 
-Int32 parseShard(const String & engine_full)
+const String PARSE_SHARD_REGEX = "shard\\s*=\\s*(\\d+)";
+const String PARSE_SHARDS_REGEX = "DistributedMergeTree\\(\\s*\\d+,\\s*(\\d+),\\s*\\d+\\)";
+const String PARSE_REPLICATION_REGEX = "DistributedMergeTree\\(\\s*(\\d+),\\s*\\d+,\\s*\\d+\\)";
+
+Int32 searchIntValueByRegex(const String & regex_s, const String & str)
 {
-    /// shard = <shard_number>
-    static std::regex shard_regex("shard\\s*=\\s*(\\d+)");
+    std::regex pattern(regex_s);
```

Comment:
We can avoid building this regex every time by caching it globally ?

Reply:
static std::regex  cannot be used here，Because regex_s changes dynamically，If you consider optimization, you can consider using  string::find_first_of

Reply:
We can change the function signature to pass in a `std::regex`